### PR TITLE
WIP: SilverStripe 4 Support

### PR DESCRIPTION
**Pull request by wilr**

### CODE REVIEW

File: _config/config.yml [hunk 1]
```diff
@@ -1,10 +1,10 @@
-SiteConfig:
+SilverStripe/SiteConfig/SiteConfig:
   extensions:
-    - NewsletterSiteConfig
+    - SilverStripe\Newsletter\NewsletterSiteConfigExtension
 ContentController:
```
Comment (andrewandante) [inline _config/config.yml]: This whole file needs namespacing

File: composer.json [hunk 2]
```diff
@@ -1,27 +1,41 @@
...
+            "dev-master": "2.x-dev"
+        },
+        "expose": [
+            "client"
+        ]
+    },
+    "autoload": {
+        "psr-4": {
+            "SilverStripe\\Newsletter\\": "src/",
+            "SilverStripe\\Admin\\Tests\\": "tests/unit/"
```
Comment (andrewandante) [inline composer.json]: Is this correct?

File: src/Control/Email/NewsletterEmail.php [hunk 3]
```diff
@@ -62,7 +65,7 @@ public function __construct($newsletter, $recipient, $fakeRecipient=false)
 
         $this->populateTemplate(new ArrayData(array(
             'UnsubscribeLink' => $this->UnsubscribeLink(),
-            'SiteConfig' => DataObject::get_one('SiteConfig'),
+            'SiteConfig' => SiteConfig::get(),
```
Comment (andrewandante) [inline src/Control/Email/NewsletterEmail.php]: `->first()`?

File: src/Control/Email/NewsletterEmail.php [hunk 4]
```diff
@@ -141,6 +142,9 @@ public function Newsletter()
         return $this->newsletter;
     }
 
+    /**
```
Comment (andrewandante) [inline src/Control/Email/NewsletterEmail.php]: Add an actual docblock or delete :)

File: src/Control/Email/NewsletterEmail.php [hunk 5]
```diff
@@ -179,6 +183,9 @@ public function UnsubscribeLink()
         }
     }
 
+    /**
+     * @return
```
Comment (andrewandante) [inline src/Control/Email/NewsletterEmail.php]: Mystery return

File: src/Control/NewsletterAdmin.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            $config = $form->Fields()->first()->getConfig();
+            $config->removeComponentsByType(GridFieldDetailForm::class)
+                ->addComponents(new NewsletterGridFieldDetailForm());
+
+            $config->getComponentByType(GridFieldDataColumns::class)
+                ->setFieldCasting(array(
+                    "Content" => "HTMLText->LimitSentences",
+            ));
+        }
+        else if ($this->modelClass == Recipient::class) {
```
Comment (andrewandante) [inline src/Control/NewsletterAdmin.php]: elseif

File: src/Control/Email/NewsletterEmail.php (hunk 3 above)
Comment (andrewandante) [reply src/Control/Email/NewsletterEmail.php]: Or `current_site_config()`?

**Comment by wilr** [comment]
@andrewandante the couple of CWP sites I've inherited only really use this module for 300-400 subscribers so current site doesn't have too many performance issues. Open any issues for the changes you'd want for performance and I can look into doing as part of the upgrade work. Goal is to have a `2.0` ready in a couple of weeks. 

Are we happy to leave it in `silverstripe-archive` or I'm happy to take over the management of the module long term now that I'll be supporting it anyways, can move it to `wilr` if `silverstripe` is unsuitable :)

**Pull request merged**